File: src/main/java/me/shedaniel/linkie/CommandApi.java
package me.shedaniel.linkie;

import discord4j.core.event.domain.message.MessageCreateEvent;
import discord4j.core.object.entity.Member;
import discord4j.core.object.entity.MessageChannel;

import java.awt.*;
import java.time.Instant;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;

public class CommandApi {
    
    protected Map<String, CommandBase> commandMap;
    protected String prefix;
    
    public CommandApi(String prefix) {
        this.commandMap = new HashMap<>();
        this.prefix = prefix.toLowerCase(Locale.ROOT);
    }
    
    public CommandApi registerCommand(CommandBase command, String... l) {
        for(String ll : l)
            commandMap.put(ll.toLowerCase(), command);
        return this;
    }
    
    public String getPrefix(boolean isSpecial) {
        return isSpecial ? "!" : prefix;
    }
    
    public void onMessageCreate(MessageCreateEvent event) {
        Member member = event.getMember().orElse(null);
        if (!event.getMember().isPresent() || member.isBot() || !event.getMessage().getContent().isPresent())
            return;
        CompletableFuture.runAsync(() -> {
            String message = event.getMessage().getContent().get();
            MessageChannel channel = event.getMessage().getChannel().block();
            String prefix = getPrefix(!channel.getType().name().startsWith("GUILD_") || (event.getGuildId().isPresent() && (event.getGuildId().get().asLong() != 432055962233470986l)));
            if (message.toLowerCase(Locale.ROOT).startsWith(prefix)) {
                String content = message.substring(prefix.length());
                String split[] = content.contains(" ") ? content.split(" ") : new String[]{content}, cmd = split[0].toLowerCase(Locale.ROOT);
                String args[] = (Arrays.asList(split)).stream().skip(1).collect(Collectors.toList()).toArray(new String[0]);
                if (commandMap.containsKey(cmd))
                    try {
                        commandMap.get(cmd).execute(LinkieBot.executors, event, member, cmd, args, channel);
                    } catch (Throwable throwable) {
                        channel.createEmbed(emd -> {
                            emd.setTitle("Linkie Error");
                            emd.setColor(Color.red);
                            emd.setFooter("Requested by " + member.getUsername() + "#" + member.getDiscriminator(), member.getAvatarUrl());
                            emd.setTimestamp(Instant.now());
                            emd.addField("Error occurred while processing the command:", throwable.getClass().getSimpleName() + ": " + throwable.getLocalizedMessage(), false);
                        }).subscribe();
                    }
            }
        }, LinkieBot.executors);
    }
    
}

File: src/main/kotlin/me/shedaniel/linkie/CommandApi.kt
package me.shedaniel.linkie

import discord4j.core.`object`.entity.User
import discord4j.core.event.domain.message.MessageCreateEvent
import discord4j.core.spec.EmbedCreateSpec
import java.awt.Color
import java.time.Instant
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executors

class CommandApi(private val prefix: String) {
    private val executors = Executors.newScheduledThreadPool(64)
    private val commandMap: MutableMap<String, CommandBase> = mutableMapOf()
    internal val commands: MutableMap<CommandBase, MutableSet<String>> = mutableMapOf()

    @Suppress("UNUSED_PARAMETER")
    fun getPrefix(isSpecial: Boolean = false): String =
//            (if (isSpecial) "+" else prefix).toLowerCase()
            prefix.toLowerCase()

    fun registerCommand(command: CommandBase, vararg l: String): CommandApi {
        for (ll in l)
            commandMap[ll.toLowerCase()] = command
        commands.getOrPut(command, ::mutableSetOf).addAll(l)
        return this
    }

    fun onMessageCreate(event: MessageCreateEvent) {
        CompletableFuture.runAsync(Runnable {
            val user: User? = event.message.author.orElse(null)
            val message: String? = event.message.content.orElse(null)
            val channel = event.message.channel.block()
            if (user == null || user.isBot || message == null || channel == null)
                return@Runnable
            val prefix = getPrefix(event.guildId.orElse(null)?.asLong() == 432055962233470986L)
            if (message.toLowerCase().startsWith(prefix)) {
                val content = message.substring(prefix.length)
                val split = if (content.contains(" ")) content.split(" ").dropLastWhile(String::isEmpty).toTypedArray() else arrayOf(content)
                val cmd = split[0].toLowerCase()
                val args = split.drop(1).toTypedArray()
                if (cmd in commandMap)
                    try {
                        commandMap[cmd]!!.execute(event, user, cmd, args, channel)
                    } catch (throwable: Throwable) {
                        try {
                            channel.createEmbed { it.generateThrowable(throwable, user) }.subscribe()
                        } catch (throwable2: Throwable) {
                            throwable2.addSuppressed(throwable)
                            throwable2.printStackTrace()
                        }
                    }

            }
        }, executors)
    }

}

fun EmbedCreateSpec.generateThrowable(throwable: Throwable, user: User? = null) {
    setTitle("Linkie Error")
    setColor(Color.red)
    user?.apply { setFooter("Requested by $discriminatedName", avatarUrl) }
    setTimestamp(Instant.now())
    addField("Error occurred while processing the command:", throwable.javaClass.simpleName + ": " + throwable.localizedMessage
            .replace(System.getenv("GOOGLEAPI"), "*")
            , false)
}